indigo_app/forms/tasks.py

```diff
@@ -52,6 +55,71 @@ def clean_title(self):
 
         return title
 
+    def is_valid(self):
+        return super().is_valid() and self.input_file_form.is_valid()
+
+    def has_changed(self):
+        return super().has_changed() or self.input_file_form.has_changed()
+
+    def save(self, commit=True):
+        task = super().save(commit=commit)
+        if commit and self.input_file_form.has_changed():
+            self.input_file_form.save()
+            task.input_file = self.input_file_form.instance
+            task.save()
+        return task
+
+
+class TaskFileForm(forms.ModelForm):
+    class Meta:
+        model = TaskFile
+        fields = ('url', 'file')
+
+    url = forms.URLField(required=False)
+    file = forms.FileField(required=False)
+    clear = forms.BooleanField(required=False)
+
+    def __init__(self, task, *args, **kwargs):
+        super().__init__(*args, **kwargs)
+        self.task = task
+
+    def save(self, commit=True):
+        file = self.cleaned_data['file']
+        url = self.cleaned_data['url']
+        task_file = self.instance
+        # save only gets called if there's a change, so we're either updating or deleting the TaskFile
+        # either way, we should delete the actual file
+        if self.initial.get('file'):
+            self.initial.get('file').delete()
+        if url:
+            resp = requests.get(url, timeout=10)
+            resp.raise_for_status()
+            filename = urlparse(url).path.split('/')[-1] or url
+            task_file.url = url
+            task_file.size = len(resp.content)
+            task_file.filename = filename
+            task_file.mime_type = resp.headers['Content-Type']
+            if self.prefix == 'input_file':
```

Comment:
it's needed for save() just because of where we store literal files — I'm moving the save() out of here and onto the task form, let me know what you think